### ISSUE TITLE
About gaussian coverage

### ISSUE THREAD

**Issue by djx99** [Xinjie-Q/GaussianImage#14]
When I plot out the Gaussian, I notice that there are some blank areas, but the pixels in those blank areas actually have values.

![image](https://github.com/user-attachments/assets/68a878a9-60cb-4c7d-b782-10806b7a5997)
![image](https://github.com/user-attachments/assets/2b1dccac-0429-46d6-a160-a555069dc94c)

My understanding is that instead of directly calculating whether the Gaussian (3* sigma) covers a single pixel, it calculates whether the outer circle of the Gaussian covers the tiled area where the pixel is located. That is, the pixel is equally likely to be affected by the Gaussian outside of its 3* sigma?
Is my understanding correct, and is this simplified and accelerated approach common in 3DGS tasks?

**Comment by lgupta-mle** [comment]
What do you mean by the "outer circle of the Gaussian"? If you look at the rasterization CUDA kernel: 

```
// process gaussians in the current batch for this pixel
        int batch_size = min(BLOCK_SIZE, range.y - batch_start);
        for (int t = 0; (t < batch_size) && !done; ++t) {
            const float3 conic = conic_batch[t];
            const float3 xy_opac = xy_opacity_batch[t];
            const float opac = xy_opac.z;
            const float2 delta = {xy_opac.x - px, xy_opac.y - py};
            const float sigma = 0.5f * (conic.x * delta.x * delta.x +
                                        conic.z * delta.y * delta.y) +
                                conic.y * delta.x * delta.y;
            const float alpha = min(1.f, opac * __expf(-sigma));
            if (sigma < 0.f || alpha < 1.f / 255.f) {
                continue;
            }

            int32_t g = id_batch[t];
            const float vis = alpha;
            const float3 c = colors[g];
            pix_out.x = pix_out.x + c.x * vis;
            pix_out.y = pix_out.y + c.y * vis;
            pix_out.z = pix_out.z + c.z * vis;
            // T = next_T;
            cur_idx = batch_start + t;
        }
```

The calculated sigma takes into account both covariance and the pixel distance from gaussian. If the sigma is too large, exp(-sigma) will be very small and consequently `alpha` will be small. They have a condition which says if `alpha < 1/255` , they don't add that gaussian's color to the pixel's render. 

Whether or not this condition on alpha is valid outside the 3*sigma range for the gaussian, I am not sure. I would like to know that as well! :)

**Comment by jzliu02** [comment]
How do you plot these gaussians? You wirte this visualizing code or it already exists somewhere in this project?

**Comment by djx99** [comment]
no, I just use matplotlib.patches.Ellipse through trained scaling and rotation.
![image](https://github.com/user-attachments/assets/2a86c835-6040-4a63-8f3a-e89ec964a72a)
![image](https://github.com/user-attachments/assets/329c108a-9d1b-4901-95d6-b421d2d5c4ed)
![image](https://github.com/user-attachments/assets/95efef61-87f1-43f5-86ac-6fd69e63eeb7)
